### PR TITLE
Enable PagedAttention through Pallas

### PR DESCRIPTION
**Pull request by wonjoolee95**
Enable PagedAttention through Pallas

Test plan:
```
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper
.
----------------------------------------------------------------------
Ran 1 test in 2.209s

OK
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper_with_dynamo
.
----------------------------------------------------------------------
Ran 1 test in 2.114s

OK
```

Todo as follow-ups:
- Add unit test for Dynamo
- Enable all other parameters for jax.experimental.pallas.ops.tpu.paged_attention.paged_attention_kernel

### CODE REVIEW

File: torch_xla/experimental/custom_kernel.py
Comment (JackCaoG) [inline torch_xla/experimental/custom_kernel.py:395]: is this a typo? It still call flash_attnetion

Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py]: Yep a typo haha. Updated.

File: torch_xla/experimental/custom_kernel.py
Comment (miladm) [inline torch_xla/experimental/custom_kernel.py:334]: how do you test this method? I see you are testing the functionality directly - via bypassing this method.

Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py:342]: Yep, we'll have two separate unit tests: one with directly bypassing this method and another one to test this wrapper. I updated PR with the latter.

**Comment by miladm** [comment]
cc @WoosukKwon

**Comment by wonjoolee95** [comment]
Locally, the tests are succeeding on my v4:
```
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper
.
----------------------------------------------------------------------
Ran 1 test in 2.209s

OK
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper_with_dynamo
.
----------------------------------------------------------------------
Ran 1 test in 2.114s

OK
```

I also just triggered the TPU CI on this PR.

**Comment by wonjoolee95** [comment]
The CPU CI is failing with an unrelated test:
```
======================================================================
FAIL: test_resnet18 (__main__.DynamoTrainingBasicTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/xla/xla/pytorch/xla/test/dynamo/test_dynamo.py", line 494, in test_resnet18
    self.assertEqual(met.metric_data('ExecuteTime')[0], sample_count * 3)
AssertionError: 29 != 30
```
The CI including the TPU CI is passing, so this PR should be good to be reviewed. A gentle ping, cc @alanwaketan 😄  Thanks!

**Review by miladm — APPROVED** [review]
Thanks @wonjoolee95  - left a comment for you to eval and address - approving to unblock you

File: torch_xla/experimental/custom_kernel.py
Comment (alanwaketan) [inline torch_xla/experimental/custom_kernel.py:334]: The original kernel has this thing called: q_dtype_for_kernel_launch? What does it do? Should we copy that as well?

File: torch_xla/experimental/custom_kernel.py
Comment (alanwaketan) [inline torch_xla/experimental/custom_kernel.py:431]: It feels like this part can be consolidated with the flash attention one.

File: test/test_pallas.py
Comment (alanwaketan) [inline test/test_pallas.py:511]: It's interesting that you use jax as the reference. I guess that works too. Wondering if we can just the eager attention helper in the class instead? Or that doesn't work? Anyway, if you are using jax as the reference, you can drop this.

File: test/test_pallas.py
Comment (alanwaketan) [inline test/test_pallas.py:543]: Can you explain what these seq_lens are? Are these the previous tokens for each batch in k, v?

**Review by alanwaketan — APPROVED** [review]
In general, it looks good to me. Left a few comments.

File: test/test_pallas.py
Comment (wonjoolee95) [reply test/test_pallas.py]: Sg, yeah I saw that we're dependent on JAX Pallas anyways, so I thought it may be easier to just test against the JAX's outputs. 

Ah, makes sense removed the `jax.config` updates.

File: test/test_pallas.py
Comment (wonjoolee95) [reply test/test_pallas.py]: Sg, updated.

File: torch_xla/experimental/custom_kernel.py
Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py]: Sg, refactored these into a helper function.

File: test/test_pallas.py
Comment (wonjoolee95) [reply test/test_pallas.py:519]: Yep, that is my understanding -- the `seq_lens` here equals the number of tokens that are processed in the batch. Reference: https://docs.vllm.ai/en/latest/dev/kernel/paged_attention.html#concepts.

File: torch_xla/experimental/custom_kernel.py
Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py:374]: In the original kernel, the `q_dtype_for_kernel_launch` is always either `jnp.float32` or q's `dtype`. In our case, I'm expecting the passed-in q's `dtype` to be `torch.32`, so the `q_dtype_for_kernel_launch` will always be `float32`.

Comment (alanwaketan) [reply torch_xla/experimental/custom_kernel.py:374]: No, I don't think that will be the case for actual workflow. It could be bf16 or even in8 etc...

Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py:374]: I see, makes sense. Just updated to handle `q_dtype_for_kernel_launch`, following jax's kernel -- https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/paged_attention/paged_attention_kernel.py#L393. I can follow-up in another PR to add some more unit tests for different dtypes for `q`.

File: torch_xla/experimental/custom_kernel.py
Comment (JackCaoG) [reply torch_xla/experimental/custom_kernel.py:427]: actually it is not even `paged attention`, you can just make this warning message more general.

File: torch_xla/experimental/custom_kernel.py
Comment (alanwaketan) [inline torch_xla/experimental/custom_kernel.py:404]: I guess you can combine this with the above L396 code.

File: torch_xla/experimental/custom_kernel.py
Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py]: Good catch, updated to use an `f` string.

**Comment by wonjoolee95** [comment]
Thanks all for the reviews. After addressing all the comments, the two unit tests are still passing locally on my V4. I'll let the TPU CI verify one more time before merging.

File: torch_xla/experimental/custom_kernel.py
Comment (alanwaketan) [inline torch_xla/experimental/custom_kernel.py:417]: You probably want to use `.to` to cast the output back to the original dtype here.

Comment (wonjoolee95) [reply torch_xla/experimental/custom_kernel.py]: Updated.

**Comment by wonjoolee95** [comment]
Merging as all CI is green.

**Pull request merged**
